external/llvm-project/mlir/lib/Conversion/AsyncToGPU/AsyncToGPU.cpp

```diff
@@ -312,30 +316,23 @@ void ConvertAsyncToGPUPass::runOnOperation() {
   ModuleOp module = getOperation();
   MLIRContext *ctx = module->getContext();
 
-  // Convert async dialect types and operations to LLVM dialect.
-  AsyncGPUTypeConverter converter;
-  RewritePatternSet patterns(ctx);
+  {
+    // Convert async.launch to gpu.launch if async.targets[gpu] exists
+    RewritePatternSet patterns(ctx);
+    patterns.add<LaunchRewritePattern>(ctx);
 
-  patterns.add<LaunchOpConversion, AwaitOpConversion>(converter, ctx);
-
-  ConversionTarget target(*ctx);
-  target.addLegalOp<arith::ConstantOp, func::ConstantOp,
-                    UnrealizedConversionCastOp>();
-  target.addLegalDialect<gpu::GPUDialect>();
-
-  // All operations from Async dialect must be lowered to the GPU dialect.
-  target.addIllegalDialect<async::AsyncDialect>();
+    if (failed(applyPatternsAndFoldGreedily(module, std::move(patterns))))
+      signalPassFailure();
+  }
 
-  // Except when async.launch has no GPU target.
-  target.addDynamicallyLegalOp<async::LaunchOp>(
-      [&](async::LaunchOp op) { return !getGPUTarget(op).has_value(); });
-  // TODO(sjw): Make async.token universal
-  // target.addDynamicallyLegalOp<async::AwaitOp>([&](async::AwaitOp op) {
-  //     return true;
-  // });
+  {
+    // Convert async.await to gpu.wait if has gpu.tokens
+    RewritePatternSet patterns(ctx);
+    patterns.add<AwaitRewritePattern>(ctx);
```

Comment:
Was it not possible to rely on the launch getting rewritten before the wait?

Reply:
I think the pattern rewriter runs in reverse order within a block so it would hit the await ops first.